### ISSUE TITLE
API after the quadrature rule struct members are private

### ISSUE THREAD

**Issue by JSorngard** [DomiDre/gauss-quad#29]
Implementing the last point of #27 requires changing the API. So I start this issue for thinking about how the API could look after the change.

## My first draft suggestion:  
While thinking of this I had the thought that we could change the quadrature rule structs to store the nodes and weights in the same allocation:

```rust
pub type Node = f64;
pub type Weight = f64;
pub type NodeWeightPair = (Node, Weight);

// or possibly
// #[derive(Debug, Clone, Copy, PartialEq)]
// pub struct NodeWeightPair {
//     node: f64,
//     weight: f64,
// }
// with additional impls for convenience functions and traits, like `pub const fn node(&self) -> f64 {&self.node}`
// and maybe `PartialEq<(f64, f64)>`.

// QuadratureRule is a placeholder for e.g. GaussLaguerre.
#[derive(Debug, Clone, PartialEq)]
pub struct QuadratureRule {
    node_weight_pairs: Vec<NodeWeightPair>,
    ...
}
```
The main benefit is that the node and its corresponding weight would then be stored together in memory, so when the CPU fetches one it gets the other in cache as well, which could in theory speed up integration. The biggest con is that I think we would need one more temporary allocation in the `new` constructor for most of the quadrature rules.  
This change would help if we ever introduce a parallelism feature, since I found rayon's `par_bridge` method to bottleneck and be basically serial when it adapts a `vec_a.iter().zip(vec_b.iter())` iterator into a parallel one (maybe there's a way to make it work, but I don't know it). If we store the data this way my suggestion for the API would be something like this:

```rust
impl QuadratureRule {
    pub fn new(input_params) -> Self {...}
    pub fn integrate(&self, interval_bounds, integrand) -> f64 {...}
    pub fn as_node_weight_pairs(&self) -> &[NodeWeightPair] {...}
    pub fn into_node_weight_pairs(self) -> Vec<NodeWeightPair> {...}
    pub fn iter(&self) -> QuadratureRuleIter<'_> {...}
    pub fn nodes(&self) -> QuadratureRuleNodes<'_> {...}
    pub fn weights(&self) -> QuadratureRuleWeights<'_> {...}
}

pub struct QuadratureRuleIter<'a> {...}
impl<'a> Iterator for QuadratureRuleIter<'a> {type Item = &NodeWeightPair; ...}
impl<'a> ExactSizeIterator for QuadratureRuleIter<'a> {...}
impl<'a> FusedIterator for QuadratureRuleIter<'a> {}
impl<'a> DoubleEndedIterator for QuadratureRuleIter<'a> {...}
// Same for QuadratureRuleNodes and QuadratureRuleWeights

pub struct QuadratureRuleIntoIter {...}
impl IntoIterator<IntoIter = QuadratureRuleIntoIter, Item = NodeWeightPair> for QuadratureRule {...}
// Same impls as for QuadratureRuleIter
```
If so the iterator structs should probably be defined in a submodule, e.g. `gauss_quad::legendre::iterators` and `gauss_quad::hermite::iterators` in order to not clog up the docs page. If it seems too wasteful to define essentially the same iterator once per module it might be nicer to either use a macro or define a `NodeWeightPairsIter` and `NodeWeightPairsIntoIter` in `lib.rs` and have the `iter` and `into_iter` functions just return an instance of those. But then the crate user would get less help from the type system.

**Comment by DomiDre** [comment]
I think the first thought to separate nodes and weights as field was that the "simple" quadrature rules like midpoint / simpson dont have varying weights. But I suppose if it enables / simplifies parallelizing that's worth a change... crate is called Gaussian quadrature in the end anyway :-).

**Comment by JSorngard** [comment]
Yeah I agree, it's slightly ugly that some of the rules wouldn't perfectly fit the pattern :/  
If we were to go this direction then I guess we would do those as

```rust
// Maybe not even this, just use the raw f64?
pub type Node = f64;
pub struct QuadretureRule {
    nodes: Vec<Node>
}
impl QuadratureRule {
    // new and integrate fns
    pub fn as_nodes(&self) -> &[Node] {...}
    pub fn into_nodes(self) -> Vec<Node> {...}
    pub fn iter(&self) -> QuadratureRuleIter<'a> {...}
}
// impl for iterator
```

**Comment by JSorngard** [comment]
I also don't want to come of as too sure that this is an obviously great idea, it could make sense to keep nodes and weights separate and only join them together if we find that it's needed in the future.

**Comment by DomiDre** [comment]
Tbh I never tried to run the integration in a parallelized manner in Rust as far as I can remember. But I suppose that is definitely something one should be able to do to integrate as fast as possible. You think it's possible to do it if nodes and weights are separated?

**Comment by JSorngard** [comment]
I'm not sure, but this is something we can check. I'll create a test branch and benchmark the different approaches.

**Comment by JSorngard** [comment]
I have added a large number of benchmarks of various sizes of Gauss-Legendre quadrature for the combination of split or unified nodes and weights as well as serial or parallel integration and simple or complex integrands.

The (too many) benchmarks can be run by checking out the `benchmark_parallel_memory_layout` branch and running `cargo bench`. The results can be viewed by opening `target/criterion/reports/index.html`.

I think it can be seen that the memory layout makes almost no difference in the serial case. It is slightly faster which is nice, but it's not by much.
It makes a huge difference in the parallel case, but the parallel case is not faster than the serial case until massive degrees.

**Comment by DomiDre** [comment]
Nice job with the benchmarks!!! 

I ran them and see the same thing in the serial and parallel case when comparing split and unified, but what do you mean with that the parallel case is not faster than the serial case? When I compare for example exp(x)sin(x)cos(x)ln(x)/unified with degree 10 for serial I get as mean +- std: 139 +- 4 micros and for parallel 17 +- 1 micros on a machine with 32 cores. I suppose the effect is more pronounced the more complex the integrand is.

So this benchmark shows that we should definitely go the way of unifying nodes & weights + enable parallelization

**Comment by JSorngard** [comment]
I'm gonna recheck as soon as I can to see if I missed something completely :O

**Comment by JSorngard** [comment]
I get that the serial unified degree 10 benchmark with the complicated integrand finishes in 163.04 ns on my CPU, while the parallel one finishes in 4.8391 µs (note the different units). That is, I get that the parallel benchmark is 30 times slower than the serial one :S

**Comment by DomiDre** [comment]
Omg somebody should take my degree away. Yes I didnt look at the unit ... 

I guess the overhead for parallelization is too much then. Now I also see that it only switches at very high degrees. 

But my gut would say if the integrand maybe was something more complicated (lets say some double integral where the inner integral is then calculated in serial and the outer in parallel), the benefits could also be obtained at lower degrees :thinking:

**Comment by JSorngard** [comment]
That makes sense, yeah. Since unifying the memory doesn't hurt anything and is potentially useful for parallelization I agree that we do it I think.

**Comment by JSorngard** [comment]
The unified memory API was implemented in #46.
